### ISSUE TITLE
Add CTS test for early-reject mapAsync

### ISSUE THREAD

**Issue by takahirox** [gpuweb/cts#2009]
A CTS test for early-reject mapAsync() should be added.

https://github.com/gpuweb/gpuweb/pull/3348

**Comment by kainino0x** [comment]
I rechecked the spec and it reminds me, immediate rejection only happens if a map is pending. We should also test that the promise *isn't* already-rejected if the buffer was already mapped. That is, the rejection should occur in a later task instead of in a microtask.

This is probably just a matter of adding a check to `testMapAsyncCall` that, in a microtask, the rejection hasn't happened yet. This can be a separate PR from #2033.

**Comment by takahirox** [comment]
> immediate rejection only happens if a map is pending. We should also test that the promise isn't already-rejected if the buffer was already mapped. That is, the rejection should occur in a later task instead of in a microtask.

Good catch, thanks.

> This is probably just a matter of (conditially) adding a check to testMapAsyncCall that, in a microtask, the rejection hasn't happened yet. This can be a separate PR from https://github.com/gpuweb/cts/pull/2033.

I made #2052. With the change [this test](https://github.com/gpuweb/cts/blob/fdc34f66b091435efbf2902a0034ee0ac0f8e0cc/src/webgpu/api/validation/buffer/mapping.spec.ts#L176) will check that `mapAsync()` doesn't reject immediately on buffers already mapped.

**Comment by takahirox** [comment]
I noticed that early-rejection test needs to be updated because pending map error is not a validation error although the current test expects pending map error as a validation error. Please refer to https://bugs.chromium.org/p/dawn/issues/detail?id=1613

I will make a PR.

**Comment by takahirox** [comment]
I also noticed that these tests need to be updated
* webgpu:api,operation,buffers,map_oom:mapAsync
* webgpu:api,validation,buffer,mapping:getMappedRange,state,mapped
* webgpu:api,validation,buffer,mapping:getMappedRange,state,mappedAtCreation

For example `await buffer;` needs to be inserted at [here](https://github.com/gpuweb/cts/blob/82ec26d039e19c70f6fa979bbff03b82d4b6ff7f/src/webgpu/api/operation/buffers/map_oom.spec.ts#L57).

See https://dawn-review.googlesource.com/c/dawn/+/113607/comments/09b70269_59aa698e

I will work on it after #2052 and #2080 will be merged.

**Comment by takahirox** [comment]
@kainino0x Friendly ping for the PRs

**Comment by kainino0x** [comment]
@takahirox apologies, haven't had a chance to filter through my catch-all email inbox. If you request my review on the PRs themselves then they'll get surfaced directly to my review inbox (let me know if you don't have the github permissions to do so)

**Comment by takahirox** [comment]
@kainino0x Thanks, added you to the reviewers. You seem to be busy now. Who else likely can be a appropriate reviewer in this domain if you don't have time? For example @Kangz ?

**Comment by kainino0x** [comment]
I plan to be working part time during the December holidays (already finished my holiday travel). I'll be a little slower but still trying to keep up with reviews. Let me update my status with that.

Generally yes, Corentin can help or find another reviewer.

**Comment by takahirox** [comment]
I think we can close this issue because the PR has been merged? Please double check because I worked on these while ago.

**Comment by kainino0x** [comment]
I don't remember for sure either but from a quick skim of the PR it looks like this is done. Thanks for updating!
